### ISSUE TITLE
Parallelization 

### ISSUE THREAD

**Issue by wvanbergen** [wvanbergen/request-log-analyzer#18]
Introduce parallelization to speed up parsing. Some ideas:
- Try the regexps for every line type in parallel.
- Parse multiple lines at once.

**Comment by barttenbrinke** [comment]
http://www.rubycentral.com/pickaxe/tut_threads.html

**Comment by barttenbrinke** [comment]
http://stackoverflow.com/questions/559420/how-to-download-fast-lots-of-web-pages-in-ruby-parallelizing-download

**Comment by barttenbrinke** [comment]
Tried parallelizing parse line. The thread construction used about the same time as the consecutive regexp parsing. Either that or my threads are not working :).

**Comment by barttenbrinke** [comment]
file_format.rb

```
# Parses a line by trying to parse it using every line definition in this file format
def parse_line(line, &warning_handler)
  request_data = nil

  threadpool = []
  request_data = nil

  count = 0
  self.line_definitions.each do |lt, definition|
    count = count + 1
    threadpool << Thread.new(lt) do |lt|
      match = definition.matches(line, &warning_handler)
      request_data = match if match
    end
  end

  threadpool.each { |t| t.join }

  return request_data
end
```

**Comment by barttenbrinke** [comment]
Wow, just removed a significant bug in parse_line. RLA is now about 3 times faster :)

**Comment by wvanbergen** [comment]
I tried to parse multiple lines in parallel. It works fine, but it does not yield any performance gains because of the lousy threading implementation in Ruby. It does increase performance when using JRuby though, but JRuby is way slower to begin with :-)

Because of these findings, parallelization will not be implemented in the master branch, but I saved experimental code in the `multithreaded` branch if someone is interested.

**Comment by barttenbrinke** [comment]
Parked :D

**Comment by wvanbergen** [comment]
Another idea: forking instead of threads. This should work better on UNIX platforms.

Issues to keep track of: the amount of processes / processor cores ratio, etc.

**Comment by barttenbrinke** [comment]
After talking with Eloy I think this should be parked until MacRuby, Ruby 1.9 or JRuby because they have basic threading support. You could take a look at the rvm gem. http://rvm.beginrescueend.com/
